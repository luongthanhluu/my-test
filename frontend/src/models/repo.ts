import { Card } from './card'

export interface RepoItem {
    id: string
    title: string
    cards: Card[]
}

export interface Repo {
    id?: string
    name: string
    lists?: RepoItem[]
}

export enum ListStatus {
    Open = 'Open',
    Confirmed = 'Confirmed',
    FalsePositive = 'False Positive',
    Fixed = 'Fixed',
}
