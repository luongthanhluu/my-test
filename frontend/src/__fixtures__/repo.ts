import { Repo } from 'models/repo'

export const repoMock: Repo[] = [
    {
        name: 'Repo 1',
        id: '1',
        lists: [
            { title: 'Open', cards: [], id: 'a' },
            {
                title: 'Confirmed',
                cards: [
                    { text: 'Vulnerability 3', id: 'y' },
                    { text: 'Vulnerability 4', id: 'z' },
                ],
                id: 'b',
            },
            {
                title: 'False Positive',
                cards: [
                    { text: 'Vulnerability 5', id: 'm' },
                    { text: 'Vulnerability 6', id: 'n' },
                ],
                id: 'c',
            },
            {
                title: 'Fixed',
                cards: [
                    { text: 'Vulnerability 7', id: 'o' },
                    { text: 'Vulnerability 8', id: 'p' },
                    {
                        text: 'Vulnerability 1',
                        id: 'w',
                        activities: [
                            {
                                createdAt: '2021-11-28T11:24:05.600Z',
                                description: 'moved from Open to Confirmed',
                            },
                            {
                                createdAt: '2021-11-28T11:24:24.903Z',
                                description: 'moved from Confirmed to Fixed',
                            },
                        ],
                    },
                ],
                id: 'd',
            },
        ],
    },
]
