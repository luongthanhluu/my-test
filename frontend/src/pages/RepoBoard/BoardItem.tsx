import { FC, useState } from 'react'
import { IconButton, Grid, Card, CardContent, Typography } from '@mui/material'
import AddIcon from '@mui/icons-material/AddCircle'

import { RepoItem } from 'models/repo'
import { AddWrapper } from 'pages/Home/styles'

import { VulnerabilityCard } from './VulnerabilityCard'

interface BoardItemProps {
    item: RepoItem
}

export const BoardItem: FC<BoardItemProps> = ({ item }) => {
    const [isAdding, setIsAdding] = useState(false)
    const onClickAddNew = () => {
        setIsAdding(true)
    }
    const onCancelAdd = () => {
        setIsAdding(false)
    }
    return (
        <Grid item xs={12} md={6} lg={3}>
            <Card variant="outlined">
                <CardContent>
                    <Typography variant="h5" component="div">
                        {item?.title}
                    </Typography>
                    {item?.cards?.map((card) => (
                        <VulnerabilityCard
                            item={card}
                            status={item?.title}
                            key={card?.id}
                        />
                    ))}
                    {isAdding && (
                        <VulnerabilityCard
                            key="adding-item"
                            onCancel={onCancelAdd}
                            listId={item.id}
                        />
                    )}
                    <AddWrapper>
                        <IconButton onClick={onClickAddNew}>
                            <AddIcon color="secondary" fontSize="large" />
                        </IconButton>
                    </AddWrapper>
                </CardContent>
            </Card>
        </Grid>
    )
}
