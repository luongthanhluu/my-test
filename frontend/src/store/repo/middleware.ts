import { createAsyncThunk } from '@reduxjs/toolkit'
import { Repo } from 'models/repo'
import {
    getListRepoApi,
    editRepoApi,
    addRepoApi,
    deleteRepoApi,
} from './../../services/repoApi'

export const repoList = createAsyncThunk('repo/repoList', async () => {
    const response = await getListRepoApi()
    return response
})

export const updateRepo = createAsyncThunk(
    'repo/updateRepo',
    async (repo: Repo) => {
        const response = await editRepoApi(repo)
        return response
    }
)

export const addRepo = createAsyncThunk('repo/addRepo', async (repo: Repo) => {
    const response = await addRepoApi(repo)
    return response
})

export const deleteRepo = createAsyncThunk(
    'repo/deleteRepo',
    async (id: string) => {
        const response = await deleteRepoApi(id)
        return response
    }
)
