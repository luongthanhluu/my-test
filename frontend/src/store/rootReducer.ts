import { combineReducers } from '@reduxjs/toolkit'

import { repoSlice } from './repo'
import { boardSlice } from './board'

export const rootReducer = combineReducers({
    repo: repoSlice.reducer,
    board: boardSlice.reducer,
})

export type RootState = ReturnType<typeof rootReducer>
