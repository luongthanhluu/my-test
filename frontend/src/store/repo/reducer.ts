import { createSlice, current, PayloadAction } from '@reduxjs/toolkit'
import {
    ERROR_ADD_REPO_MESSAGE,
    ERROR_DELETE_REPO_MESSAGE,
    ERROR_UPDATE_REPO_MESSAGE,
    ERROR_LOAD_REPO_MESSAGE,
} from 'constants/message'
import { Repo } from 'models/repo'
import { getIndexRepoById } from 'utils/repoHelper'
import { repoList, updateRepo, addRepo, deleteRepo } from './middleware'

export interface RepoState {
    list: {
        loading: boolean
        data: Repo[]
        errorMessage: string
    }
}

const initialState: RepoState = {
    list: {
        loading: false,
        data: [],
        errorMessage: '',
    },
}

// Ref: https://redux-toolkit.js.org/tutorials/basic-tutorial
// Slice can create actions and reducer together
export const repoSlice = createSlice({
    name: 'repo',
    initialState: initialState,
    reducers: {},
    extraReducers: {
        //get list repo
        [repoList.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [repoList.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo[]>
        ) => {
            state.list.loading = false
            state.list.data = payload
        },
        [repoList.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_LOAD_REPO_MESSAGE
        },

        //update repo name
        [updateRepo.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [updateRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo>
        ) => {
            state.list.loading = false
            const listRepo = current(state.list.data)
            const index = getIndexRepoById(listRepo, payload?.id || '')
            if (index > -1) {
                state.list.data[index] = payload
            }
        },
        [updateRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_UPDATE_REPO_MESSAGE
        },

        // add new repo
        [addRepo.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [addRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo>
        ) => {
            state.list.loading = false
            state.list.data = [...state.list.data, payload]
        },
        [addRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_ADD_REPO_MESSAGE
        },

        // delete new repo
        [deleteRepo.pending.toString()]: (state) => {
            state.list.loading = true
            state.list.errorMessage = ''
        },
        [deleteRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<string>
        ) => {
            state.list.loading = false
            const listRepo = current(state.list.data)
            const index = getIndexRepoById(listRepo, payload || '')
            if (index > -1) {
                const newList = [...listRepo]
                newList.splice(index, 1)
                state.list.data = newList
            }
        },
        [deleteRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_DELETE_REPO_MESSAGE
        },
    },
})
