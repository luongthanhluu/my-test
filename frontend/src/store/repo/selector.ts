import { RootState } from '../rootReducer'

export const getListRepoData = (state: RootState) => state.repo?.list
