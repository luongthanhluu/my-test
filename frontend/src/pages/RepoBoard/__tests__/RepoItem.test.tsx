import React from 'react'
import { render } from 'enzyme'
import { Provider } from 'react-redux'

import { store } from 'store'
import { BoardItem } from '../BoardItem'
import { repoMock } from '../../../__fixtures__/repo'

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: jest.fn(),
}))

test('should match snapshot with data', () => {
    const wrapper = render(
        <Provider store={store}>
            <BoardItem item={repoMock[0].lists[0]} />
        </Provider>
    )
    expect(wrapper).toMatchSnapshot()
})
