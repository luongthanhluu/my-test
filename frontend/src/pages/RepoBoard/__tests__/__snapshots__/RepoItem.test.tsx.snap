// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with data 1`] = `
<div
  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-3 css-17jswnn-MuiGrid-root"
>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Open
      </div>
      <div
        class="MuiBox-root css-15z49vq"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge css-1nryff-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
